lib/spack/spack/test/conftest.py

```diff
@@ -884,6 +886,126 @@ def mock_archive(request, tmpdir_factory):
         expanded_archive_basedir=spack.stage._source_path_subdir)
 
 
+@pytest.fixture(scope='session')
+def mock_cvs_repository(tmpdir_factory):
+    """Creates a very simple CVS repository with two commits and a branch."""
+    cvs = spack.util.executable.which('cvs', required=True)
+
+    tmpdir = tmpdir_factory.mktemp('mock-cvs-repo-dir')
+    tmpdir.ensure(spack.stage._source_path_subdir, dir=True)
+    repodir = tmpdir.join(spack.stage._source_path_subdir)
+    cvsroot = str(repodir)
+
+    # The CVS repository and source tree need to live in a different directories
+    sourcedirparent = tmpdir_factory.mktemp('mock-cvs-source-dir')
+    module = spack.stage._source_path_subdir
+    url = cvsroot + "%module=" + module
+    sourcedirparent.ensure(module, dir=True)
+    sourcedir = sourcedirparent.join(module)
+
+    def format_date(date):
+        return date.__format__('%Y-%m-%d %H:%M:%S')
+    revision_date = {}
+
+    # Initialize the repository
+    with sourcedir.as_cwd():
+        cvs('-d', cvsroot, 'init')
+        cvs('-d', cvsroot, 'import', '-m', 'initial mock repo commit',
+            module, 'mockvendor', 'mockrelease')
+        with sourcedirparent.as_cwd():
+            cvs('-d', cvsroot, 'checkout', module)
+
+        # Commit file r0
+        r0_file = 'r0_file'
+        sourcedir.ensure(r0_file)
+        cvs('-d', cvsroot, 'add', r0_file)
+        cvs('-d', cvsroot, 'commit', '-m', 'revision 0', r0_file)
+        revision_date['1.1'] = format_date(datetime.now())
```

Comment:
For this, I need to call `cvs` with that made-up time. I assume `cvs` queries the kernel to find the current date and time. Is this possible in Spack's mocking infrastructure?

Reply:
Now that I read through this more, are these generated dates actually used? It looks like they are a fallback when `get_date` should be pulling a date from `cvs`? If the act of creating `cvs` commits records timestamps, then I see why we would be working with real times, but I don't see the reason for recording these other times outside of `cvs`.

We cannot mock anything that `cvs` calls (we are only able to update the behavior of python functions). Actually strictly speaking that's not true since if it calls an exe to get this, we could override `PATH`, but I'd prefer to avoid that.

Reply:
This is the code that creates a test repo. I record the times here so that I can use these same dates later in the test case to check out the repo at this date (to check checking out by date).

Reply:
Does this depend on the timestamps in the repo matching the recorded times? I think it's possible they could differ (even though they run right after one another). Could you retrieve the date from the repo as you add revisions?

Reply:
No, the time stamps do not need to match. Yes, I expect them to differ from time to time, in particular on a slow file system.

When you specify a time, then cvs will look at the state of the repo at that time, i.e. it will (for each file separately) use the latest commit _not later_ than the time you specify. Thus choosing a time that is later than the latest commit will work fine.

In a realistic use case one would specify "the repo state on June 8th", not the exact date or time of the latest commit.

Reply:
OK I think I understand now. Can you add a comment explaining this e.g. to the declaration of `revision_date`:

```
# We use this to record the times when we create CVS revisions, so that we can later
# check that we retrieve the proper revision when specifying a date (CVS guarantees
# checking out the lastest revision before or on the specified date). As we create each
# revision, we separately record the time (we expect that for a sequence of CVS
# invocations and datetime.now() calls, that the resulting timestamps have the same
# ordering)
revision_date = {}
```

Unfortunately several things might screw up that last assumption. However, if there is no way to ask CVS "what is the date of the latest revision you have" then using `datetime.now` seems like an appropriate fallback (still, I want a warning there if the CI environment changes in a way that violates this assumption).

Reply:
> if there is no way to ask CVS "what is the date of the latest revision you have" then using datetime.now seems like an appropriate fallback 

I see at https://stackoverflow.com/questions/1044651/how-to-programatically-get-the-latest-commit-date-on-a-cvs-checkout there appears to be a means to retrieve the latest timestamp for a specific file. If there are problems with this test in the future related to timestamp ordering, then I think that link would be useful (although I don't think this needs to be done now).

Reply:
I added a respective note in a comment. Thanks.